File: core/src/commonMain/kotlin/work/socialhub/kslack/api/methods/SlackApiException.java
package com.github.seratch.jslack.api.methods;

import com.github.seratch.jslack.SlackConfig;
import com.github.seratch.jslack.common.json.GsonFactory;
import net.socialhub.http.HttpResponse;
import net.socialhub.logger.Logger;

public class SlackApiException extends Exception {

    private static final Logger log = Logger.getLogger(SlackApiException.class);

    private final HttpResponse response;
    private final String responseBody;
    private final SlackApiErrorResponse error;

    public SlackApiException(HttpResponse response, String responseBody) {
        this(SlackConfig.DEFAULT, response, responseBody);
    }

    public SlackApiException(SlackConfig config, HttpResponse response, String responseBody) {
        this.response = response;
        this.responseBody = responseBody;
        SlackApiErrorResponse parsedErrorResponse = null;
        try {
            parsedErrorResponse = GsonFactory.createSnakeCase(config).fromJson(responseBody, SlackApiErrorResponse.class);
        } catch (Exception e) {
            if (log.isDebugEnabled()) {
                String responseToPrint = responseBody.length() > 1000 ? responseBody.subSequence(0, 1000) + " ..." : responseBody;
                log.debug("Failed to parse the error response body: " + responseToPrint);
            }
        }
        this.error = parsedErrorResponse;
    }

    public HttpResponse getResponse() {
        return this.response;
    }

    public String getResponseBody() {
        return this.responseBody;
    }

    public SlackApiErrorResponse getError() {
        return this.error;
    }
}

File: core/src/commonMain/kotlin/work/socialhub/kslack/api/methods/SlackApiException.kt
package work.socialhub.kslack.api.methods

import work.socialhub.khttpclient.HttpResponse
import work.socialhub.kslack.api.methods.helper.JsonHelper.fromJson

/**
 * Exception thrown when a Slack Web API call results in an HTTP error (non-2xx status code).
 *
 * This exception wraps the raw [HttpResponse] and attempts to parse the response body
 * as a [SlackApiErrorResponse] for structured error information. The original HTTP
 * status code and response body are accessible via [response].
 *
 * Example:
 * ```kotlin
 * try {
 *     slack.chat().chatPostMessageBlocking(ChatPostMessageRequest(...))
 * } catch (e: SlackApiException) {
 *     val status = e.response.statusCode
 *     val errorBody = e.error?.error
 * }
 * ```
 *
 * @see HttpResponse
 * @see SlackApiResponse
 */
class SlackApiException(
    /**
     * The raw HTTP response from the Slack API.
     * Contains status code, headers, and body.
     */
    val response: HttpResponse,
) : Exception() {

    /**
     * Parsed error response from the Slack API body.
     * Contains structured error information like error code, required scopes, etc.
     * May be null if the response body cannot be parsed as JSON.
     */
    var error: SlackApiErrorResponse? = null

    init {
        val body = response.stringBody
        try {
            error = fromJson(body)

        } catch (e: Exception) {
            val text = if (body.length > 1000)
                body.subSequence(0, 1000).toString() + " ..." else body
            println("Failed to parse the error response body: $text")
        }
    }

}
